File: Template/String/AhoCorasick.cpp
/*
Aho-Corasick Automaton (AC 自動機)

Problems:
- https://www.luogu.com.cn/problem/P3808
*/

#include <bits/stdc++.h>
using namespace std;
constexpr int ALPH = 26;
#define endl '\n'

struct Node {
    array<Node*, ALPH> child;
    Node* fail;
    int cnt;
    Node() : fail(nullptr), cnt(0) {
        fill(child.begin(), child.end(), nullptr);
    }
};

class AhoCorasick {
public:
    Node* root;

    AhoCorasick() {
        root = new Node();
    }

    void insert(const string& word) {
        Node* node = root;
        for (char ch : word) {
            int c = ch - 'a';
            if (node->child[c] == nullptr) {
                node->child[c] = new Node();
            }
            node = node->child[c];
        }
        node->cnt += 1;  // 累加，防止有相同的模式串
    }

    void build() {
        root->fail = root;
        // BFS
        queue<Node*> q;
        for (int i = 0; i < ALPH; ++i) {
            if (root->child[i] == nullptr) {
                root->child[i] = root;
            } else {
                root->child[i]->fail = root;
                q.push(root->child[i]);
            }
        }
        while (!q.empty()) {
            Node* u = q.front();
            q.pop();
            for (int i = 0; i < ALPH; ++i) {
                if (u->child[i] == nullptr) {
                    u->child[i] = u->fail->child[i];
                } else {
                    Node* v = u->child[i];
                    v->fail = u->fail->child[i];
                    q.push(v);
                }
            }
        }
    }
};

void solve() {
    int n;
    cin >> n;
    vector<string> words(n);
    for (int i = 0; i < n; ++i) cin >> words[i];
    string t;
    cin >> t;

    AhoCorasick ac;
    for (const auto& word : words) ac.insert(word);
    ac.build();

    int ans = 0;
    Node* node = ac.root;
    for (char ch : t) {
        int idx = ch - 'a';
        node = node->child[idx];  // 由於是 Trie 圖，直接轉移即可

        // 沿著 fail 鏈向上搜集所有匹配的模式串
        Node* temp = node;
        while (temp != ac.root && temp->cnt != -1) {
            ans += temp->cnt;
            temp->cnt = -1;  // 標記為 -1，代表此節點已被統計過，避免重複計算
            temp = temp->fail;
        }
    }
    cout << ans << endl;
    return;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    solve();
    return 0;
}

File: Template/String/AhoCorasick.py
from typing import List
from collections import deque

"""
Aho-Corasick Automaton
令 u.goto[c] 表示在節點 u 時，匹配字元 c 後下一個節點，方便失配時直接跳到下一個可能匹配的位置，不用從 fail 往上找。
- 如果 u.child[c] 存在，則 u.goto[c] = u.child[c] 
- 如果 u.child[c] 不存在，則 u.goto[c] = u.fail.goto[c]
這裡直接用 child 當作 goto，但會破壞原本的 Trie 結構，需要注意。

fail 指標有個性質是可以構成一棵樹，稱為 fail tree，根節點是 root，其他節點的父節點是它的 fail 指向的節點。
因為 fail 只會指向深度比自己淺的節點，所以 fail tree 是一棵樹。

Template:
- P3808 AC 自动机（简单版）
  因為只是問有沒有出現，因此只需要延 fail tree 往上找，並標記 fail tree 上訪問過的節點，避免二次訪問即可。
- P3796 AC 自动机（简单版 II）
  要統計出現次數，因為有些節點不是某個 word 的結尾，沿著 fail tree 往上找會浪費時間，
  又因為要統計出現次數不能直接標記，這時候就要使用 last 指標跳過沒有用的節點。
  另外因為要統計出現次數，本題需要保留每個 word 對應的節點。
- P5357 【模板】AC 自动机.py
  但是就算使用 last 指標，沿著 last 指標往上找最壞可以是 O(n)，其中 n 是模式串的數量，在模式串太多的情況下，也會浪費時間。
  比較好的做法是不直接每次都往上找，而是先把所有節點按照 fail tree 的拓撲順序排序，
  然後從後往前統計出現次數並累加貢獻，這樣每個節點只會被訪問一次。

Problem:
- 3213. Construct String with Minimum Cost
- 3292. Minimum Number of Valid Strings to Form Target II

Reference:
- https://leetcode.cn/problems/minimum-number-of-valid-strings-to-form-target-ii/solutions/2917929/ac-zi-dong-ji-pythonjavacgo-by-endlessch-hcqk
"""


class Node:
    def __init__(self):
        self.child = [None] * 26
        self.fail = None  # fail pointer
        self.last = None  # suffix link，用來快速跳到一定是某個 word 結尾的節點
        # self.is_end = False
        # self.depth = depth
        self.length = 0  # 可以取代 is_end 和 depth


class AhoCorasick:
    def __init__(self):
        self.root = Node()
        # self.nodes = []  # 保留每個 word 對應的節點

    def insert(self, word: str):
        node = self.root
        for ch in word:
            c = ord(ch) - ord("a")
            if not node.child[c]:
                node.child[c] = Node()
            node = node.child[c]
        node.length = len(word)
        # self.nodes.append(node)  # 保留每個 word 對應的節點

    def build(self):  # O(|Σ|N)，N 是節點數；若 |Σ|=26 視為常數，則為 O(N) = O(L)
        self.root.fail = self.root.last = self.root
        # BFS
        q = deque()
        # order = []  # 用來保存 fail tree 的拓撲順序
        for i, v in enumerate(self.root.child):
            if v is None:
                # 添加虛擬子節點
                self.root.child[i] = self.root
            else:
                v.fail = v.last = self.root
                q.append(v)
                # order.append(v)
        while q:
            u = q.popleft()
            for i, v in enumerate(u.child):
                if v is None:
                    # 添加虛擬子節點
                    u.child[i] = u.fail.child[i]
                else:
                    # 失配位置
                    v.fail = u.fail.child[i]
                    # 上一個一定是某個 word 結尾的節點
                    v.last = v.fail if v.fail.length else v.fail.last
                    q.append(v)
                    # order.append(v)
        # self.order = order

    def build2(self):  # 最壞 O(|Σ|N + L^2)；若 fail 回退很少，實務上接近 O(|Σ|N)
        self.root.fail = self.root.last = self.root
        # BFS
        q = deque()
        for v in self.root.child:
            if v is None:
                continue
            v.fail = v.last = self.root
            q.append(v)
        while q:
            u = q.popleft()
            for i, v in enumerate(u.child):
                if v is None:
                    continue
                fu = u.fail
                while fu != self.root and fu.child[i] is None:
                    fu = fu.fail
                if fu.child[i] is not None:
                    v.fail = fu.child[i]
                else:  # 如果一路找到根節點都沒有，則 fail 指向根
                    v.fail = self.root
                v.last = v.fail if v.fail.length else v.fail.last
                q.append(v)

    def traverse2(self, word: str):
        node = self.root
        for ch in word:
            idx = ord(ch) - ord("a")
            while node != self.root and node.child[idx] is None:
                node = node.fail
            if node.child[idx] is not None:
                node = node.child[idx]


class Solution3213:
    def minimumCost(self, target: str, words: List[str], costs: List[int]) -> int:
        ac = AhoCorasick()

        def insert(self, word: str, cost: int):
            node = self.root
            for ch in word:
                c = ord(ch) - ord("a")
                if not node.child[c]:
                    node.child[c] = Node()
                node = node.child[c]
            node.length = len(word)
            if not hasattr(node, "cost"):
                node.cost = cost
            else:
                node.cost = min(node.cost, cost)  # 避免相同單字有不同 cost

        ac.insert = insert.__get__(ac)  # 綁定方法到 ac 實例

        for word, cost in zip(words, costs):
            ac.insert(word, cost)
        ac.build()

        n = len(target)
        f = [0] + [float("inf")] * n
        node = ac.root
        for i, ch in enumerate(target, 1):
            node = node.child[ord(ch) - ord("a")]
            if node.length:  # 匹配到了一個盡可能長的 words[k]
                f[i] = min(f[i], f[i - node.length] + node.cost)
            # 沿著 last 往上尋找，可能可以找到其餘更短，但可以使成本更小的 words[k]
            temp = node.last
            while temp != ac.root:
                f[i] = min(f[i], f[i - temp.length] + temp.cost)
                temp = temp.last
        return f[n] if f[n] != float("inf") else -1


sol = Solution3213()
print(
    sol.minimumCost("abcdef", ["abdef", "abc", "d", "def", "ef"], [100, 1, 1, 10, 5])
)  # 7
